File: LeetCode/p216.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int sum) {
        m_k = k;
        m_sum = sum;
        backtracking(0, 0, 1);
        return result;
    }

private:
    int m_k;
    int m_sum;
    vector<int> tmp_vec;
    vector<vector<int>> result;

    void backtracking(int n, int sum, int start) {
        if (n == m_k) {
            if (sum == m_sum)
                result.emplace_back(tmp_vec);
            return;
        }
        if (sum > m_sum)
            return;
        for (int i = start; i <= 9; ++i) {
            tmp_vec.push_back(i);
            backtracking(n + 1, sum + i, i + 1);
            tmp_vec.pop_back();
        }
    }
};

File: LeetCode/p216.rs
struct Combine {
  n: i32,
  k: i32,
  tmp: Vec<i32>,
  ret: Vec<Vec<i32>>,
}

impl Combine {
  pub fn combine(mut self, k: i32, n: i32) -> Vec<Vec<i32>> {
    self.n = n;
    self.k = k;
    self.backtrack(1, 0);
    self.ret
  }

  fn backtrack(&mut self, start: i32, sum: i32) {
    if sum > self.n {
      return;
    }
    if self.tmp.len() == self.k as usize {
      if sum == self.n {
        self.ret.push(self.tmp.clone());
      }
      return;
    }
    for i in start..10 {
      self.tmp.push(i);
      self.backtrack(i + 1, sum + i);
      self.tmp.pop();
    }
  }
}

impl Solution {
  pub fn combination_sum3(k: i32, n: i32) -> Vec<Vec<i32>> {
    Combine{k, n , tmp: Vec::with_capacity(k as usize), ret: vec![]}.combine(k, n)
  }
}

